aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOAwsClientFactory.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws.s3;
+
+import java.util.Map;
+import org.apache.iceberg.aws.AwsClientFactories;
+import org.apache.iceberg.aws.AwsClientProperties;
+import org.apache.iceberg.aws.HttpClientProperties;
+import org.apache.iceberg.util.PropertyUtil;
+import software.amazon.awssdk.services.s3.S3Client;
+
+public class S3FileIOAwsClientFactory {
+  /**
+   * This property allows customizing AWS client factory implementation class. If this property is
+   * set, only {@link S3FileIOAwsClientFactory} class will be instantiated. If this property wasn't
+   * set, will load one of {@link org.apache.iceberg.aws.AwsClientFactory} factory classes to
+   * provide backward compatibility.
+   */
+  public static final String CLIENT_FACTORY = "s3.client-factory-impl";
```

Comment:
yes, my original thought is that `S3FileIOAwsClientFactory` is a trimmed down version of ` AwsClientFactories` interface, that only requires `s3()` and `s3Async()` (once we add async client integration). And there could be a `DefaultS3FileIOAwsClientFactory` that is the default implementation of that trimmed down version of the factory. So this should actually not be a boolean config, but takes the impl class name, basically exactly the same experience as the original `AwsClientFactories`

Reply:
`S3FileIOAwsClientFactory` is an interface. And, `DefaultS3FileIOAwsClientFactory`  will implement `S3FileIOAwsClientFactory`. In case the property isn’t set, do we want to fallback to `AwsClientFactory.defaultFactory()`?